typing/typecore.ml

```diff
@@ -6223,14 +6303,17 @@ and type_statement ?explanation env sexp =
     exp
   end
 
-and type_unpacks ?(in_function : (Location.t * type_expr * bool) option)
+(* Type the body within an environment extended with the unpacked modules are
+   added, returning both the typechecked body and the extended environment. *)
+and type_unpacks
```

Comment:
It looks to me that this function can be simplified, by separating out the add-module-vars-to-env functionality to a `add_module_variables` function analogous to `add_pattern_variables`. Just like `add_pattern_variables`, the new `add_module_variables` would be called toward the end of `type_pattern_list`. And it would have to be called in `type_cases`, just after the current call to `add_pattern_variables`. The refactoring seems good, because it makes module variables more like normal variables (uniformity, yay!) -- and it avoids the second call to `type_unpacks'` in `type_cases`. (Currently, the two `type_unpacks'` calls both do the work of adding module stuff to the env't, but I think we can just do this once.) Ditto the calls to `type_unpacks` in `type_let`.

Do you think this would work? If so, then we could kill off `type_unpacks` entirely, just replacing it with `type_expect`.

Reply:
As we discussed, I did this in https://github.com/ncik-roberts/ocaml-jst/pull/1, which I could have done more to get on your radar (sorry!).